### ISSUE TITLE
[Performance] PyTorch (MPS) is faster than MLX in backward of convolution layer

### ISSUE THREAD

**Issue by arnold-yan** [ml-explore/mlx#1313]
**Describe the bug**
Recently I profiled the neural network layer performance from MLX and compared with PyTorch. I found that although MLX forwarding is consistently faster than PyTorch, in some chips (M1 Pro, M1 Max), PyTorch is much faster (3x~6x) for convolution forward + backward. While in some chips such as M3 Max, MLX is faster than PyTorch.
![image](https://github.com/user-attachments/assets/43361e8d-48b5-44bc-a339-7d766de6ee1e)

**To Reproduce**
To reproduce this, I have two minimal examples. The networks just have several convolution layers. You may try these two scripts to verify the performance.

[time_pytorch_mlx.zip](https://github.com/user-attachments/files/16550493/time_pytorch_mlx.zip)

**Comment by awni** [comment]
Same benchmark on an M2 Ultra

```
average time of Pytorch: 7.20261025428772
average time of MLX: 2.34059739112854
```

**Comment by abdussamettrkr** [comment]
On M2Pro

- average time of MLX: 4.472527980804443
- average time of Pytorch: 10.073836088180542

**Comment by alwint3r** [comment]
On M3 Max

* average time of MLX: 2.4813356399536133
* average time of PyTorch: 7.1081931591033936

**Comment by awni** [comment]
Thanks for the benchmarks everyone! There is clearly an unexpected performance cliff on M1 machines here as MLX is substantially faster on M2+. We'll need to take a deeper look at that to figure out where it's coming from.

**Comment by pyvadev** [comment]
On M1
 - average time of MLX: 30.113215446472168
 - average time of Pytorch: 15.948616743087769

**Comment by jrp2014** [comment]
M3 Max: 
average time of MLX: 2.939736843109131
average time of Pytorch: 5.9829957485198975

**Comment by awni** [comment]
@arnold-yan. I took a look at this benchmark.

The performance issue turns out to be from the gradient of the second call to `nn.Upsample`. It uses nearest neighbor interpolation by default. The forward is a gather under the hood and the backward is a scatter add. The scatter add is very inefficient on M1 in this case because it uses atomics and there are a lot of collisions to the same element.

A simple fix is to use linear interpolation (which I believe is what you do with Pytorch anyway).

Changing to:

```python
        upsample = nn.Upsample(scale_factor=(h_scale, w_scale), mode="linear")
```

The benchmark runs in 2.89 ms on my M1 max compared to PyTorch 13.9 ms.

**Comment by awni** [comment]
Hi @arnold-yan https://github.com/ml-explore/mlx/pull/1541 should improve your benchmark a lot. I ran it on an M1 Max and M3 Max and the numbers are now:

Machine | MLX | PT
---- | --- | ---
M1 Max | 4.615 | 11.93 |
M3 Max | 1.938 | 10.77 |

**Comment by arnold-yan** [comment]
> Hi @arnold-yan #1541 should improve your benchmark a lot. I ran it on an M1 Max and M3 Max and the numbers are now:
> 
> Machine	MLX	PT
> M1 Max	4.615	11.93
> M3 Max	1.938	10.77

Thanks @awni! Let me test on this PR.

**Comment by arnold-yan** [comment]
> @arnold-yan. I took a look at this benchmark.
> 
> The performance issue turns out to be from the gradient of the second call to `nn.Upsample`. It uses nearest neighbor interpolation by default. The forward is a gather under the hood and the backward is a scatter add. The scatter add is very inefficient on M1 in this case because it uses atomics and there are a lot of collisions to the same element.
> 
> A simple fix is to use linear interpolation (which I believe is what you do with Pytorch anyway).
> 
> Changing to:
> 
> ```python
>         upsample = nn.Upsample(scale_factor=(h_scale, w_scale), mode="linear")
> ```
> 
> The benchmark runs in 2.89 ms on my M1 max compared to PyTorch 13.9 ms.

By the way, I also tried this change on your PR. Now the MLX time change to
```
average time of MLX: 18.393556356430054
```
It is a little bit slower than PyTorch. Does it mean we still have potential to accelerate MLX performance for linear mode?
